README
<h1>Introduction</h1>

*Fed up with how TuneIn tries to be a social network these days? How it has problems running streams even if other apps do it just fine without setting the buffer to 30 seconds or more (and then still has trouble playing the stream)? Would you like an alarm feature that actually works reliably? Would you like Chromecast [1] support with that?*

Well, I think Icerrr may be just the thing for you!

And it get's better: Icerrr is FREE [2]

Icerrr is FMAlarm's successor and currently (heavily) under development. It's built with Cordova/Phonegap and should eventually also support iOS and regular browsers. 

For now, it's just the Android app.

Expect updates. And probably a bug here and there. But you'll be helping me building this thing by using it and reporting bugs, feature requests, etc.

Join the beta program:
http://android.rejh.nl/droidapps/?app=Icerrr

[1] Chromecast support is prototype-ish.
[2] Icerrr can be downloaded for free for at least a limited period of time.

Keywords: radio, stream, icecast, shoutcast, chromecast, clock radio, alarm, cordova, phonegap, music

More info:
http://android.rejh.nl/droidapps/?app=Icerrr

<h1>Get it on Google Play</h1>

https://play.google.com/store/apps/details?id=com.rejh.icerrr.droidapp

<h1>Join the beta</h1>

Join here:
https://plus.google.com/communities/115139919019529491940

then here:
https://play.google.com/apps/testing/com.rejh.icerrr.droidapp

and then go here:
https://play.google.com/store/apps/details?id=com.rejh.icerrr.droidapp

<h1>Resources</h1>

<h2>Style + design<h2>

  * Palettes: 
    * [http://www.colourlovers.com/business/trends/branding/7880/Papeterie_Haute-Ville_Logo Papeterie_Haute-Ville_Logo]
    * [http://www.colourlovers.com/palette/3074952/Save_the_Arctic Save_the_Arctic]

  * Icon generators:
    * Material: http://shreyasachar.github.io/AndroidAssetStudio/index.html

<2h>Radio Directory<h2>

* Dirble: http://dirble.com/developer/api

= Contact =

M: icerrr@rejh.nl <br>
G: https://plus.google.com/+RoderickGadellaa <br>
T: https://twitter.com/rgadellaa <br>


androidapp/changelog.txt
0.356
  * Landscape mode for station details screen

0.355
  * Fix for importing icons
  * Minor style tweak

0.354
  * Secure communication with Icerrr API via https

0.353
  * New: On Android Wear, say "OK Google, Open radio"

0.352
  * New: control radio with intents (from apps like Tasker)

0.351
  * Fix: Issue selecting station via details screen after install

0.350
  * New: option to disable Chromecast support (frees up resources)

0.349
  * Added basic http authentication support for streams (beta, usage: 'http://user:pass@server.com')
  * Fix: streams that used any other port than 80 couldn't be manually added
  * A pinch of left over love from valentine's day

0.348
  * Fix for streams getting stuck sometimes (info code 703, 701)

0.347
  * Bugfixes

0.346
  * Fixes for stream getting stuck in buffering state
  * Fix for now playing info sometimes showing "& #4" at the end

0.344
  * Important bug fix for 0.341
  * Fix: spotify link in station details
  * Fix for manually adding stations
  * Fixes and improvements

0.339
  * Cancel 'validate station' by pressing back button

0.338
  * Fix: spotify link in station details

0.337
  * Fix for manually adding stations
  * Minor improvements

0.334
  * Minor improvements

0.333
  * Ditched polymer icons for improved performance

0.332
  * Fix: stream was resumed after call when paused by user
  * Fix: app not getting passed splash screen on older android versions (<5.0)

0.331
  * Bugfixes, improvements

0.329
  * Bugfixes, improvements

0.328
  * Migrated album art search to Google Custom Search API

0.327
  * Fix for station details not always showing station icon
  * Minor fixes for cancelling chromecast playback

0.326
  * Optimized javascript, css and images

0.325
  * Bugfixes and improvements

0.323
  * Bugfixes and improvements

0.322
  * Improved how required permissions are requested (Android 6.0+)
  * Tweaks for "Station details" screen

0.321
  * Request required permissions at startup (Android 6.0+)
  * Reduced apk size, working on slimming it down even more
  * Migrated project to Android Studio/Gradle

0.320
  * Longpressing a station now selects it
  * Tweaked swipe behaviour
  * Load polymer async

0.319
  * New: swipe between all and starred stations
  * Improvements
  
0.318
  * Critical fix for 0.317

0.317
  * Fix: choosing default icon in icon search no longer causes limbo
  * Fix: importing an icon should no longer fail

0.316
  * Enabled 'Unlock device for alarms' option (Settings)
  * Bugfixes, tweaks and improvements

0.315
  * Target SDK 23 (Android 6.0 Marshmallow)
  * Make sure alarms fire even when device is in Doze mode

0.314
  * New: report analytics (android version, app version, device model)
  * Fix: station icon got lost when editing a station

0.313
  * New UI for changing station icons when editing stations

0.312
  * Fixed #222: Import icons from Google Drive, Google Photos, etc

0.311
  * Replaced most (actionbar) icons with polymer versions! Ripples!
  * Visual tweaks for add/edit alarms and stations
  * New: import a station icon from storage, dropbox, etc
  * Prevent android media scanner from scanning Icerrr cache
  * Bugfixes and improvements

0.310
  * Fix: Service acquires wifi lock to prevent wifi from sleeping while streaming
  * Birthday edition version code

0.259
  * Pushing pixels
  * Bugfixes, improvements

0.258
  * Moved Scrobbling in settings to have its own category
  * New scrobble option: verify now playing with Echonest database
  * Option for alternative flat launcher icon (for use with Nova, Apex)
  * Fix: Spotify and Play Music search in Details searching for wrong track
  * Bugfixes, tweaks, improvements and an easter egg that should pop up any day now :)

0.257
  * New color palette is official!
  * New 'details' screen for managing a station, checking recently played, etc
  * New: Simple Last.fm Scrobbler integration (Settings, issue #207)
  * Better sync between in-app, notification and lockscreen "now playing" info
  * Lots of fixes, improvements

0.256
  * Show album art on lockscreen if already cached locally
  * Slightly but vibrantly colorize in-app artwork
  * Fix: Update selected station in app when using next/previous from notification
  * Minor fixes and improvements

0.255
  * Fix: Play button kept spinning after stopping a chromecast session
  * Fix: removing a station checks alarms and can remove them, too

0.254
  * New: backup and restore station data (in Settings)

0.253
  * More styling for timepicker
  * Stop media playback when downloading update

0.252
  * Dialogs (including timepicker) are now themed!
  * Fix: nullpointers :S

0.251

webapp/api/api.php
<?

/*

	Icerrr API

	Examples:
	- GET
		- ?a=get&q={"get":"stations"[,"last_update_time_ms":[ms]]}
		- ?a=get&q={"get":"station_info","station_id":"[station_id]"[,"last_update_time_ms":[ms]]}
		- ?a=get&q={"get":"station_nowplaying","station_id":"[station_id]"[,"last_update_time_ms":[ms]]}

	Todos:
	- Implement HTTP 304 not modified
	- Error: also mail (me)?

*/

// TODO: should only be used for debugging
error_reporting(E_ERROR | E_PARSE);

include("s.config.php");
include("s.functions.php");

cleanupjson();

$action = $_GET["a"];
$query = urldecode($_GET["q"]);
$apikey = urldecode($_GET["apikey"]);

// Checkin'
if (!$action) { error("GET['action'] is not defined"); }

// Header
header("Content-Type: application/json");
header("Access-Control-Allow-Origin: *");

// Actions
switch($action) {

	// GET
	case "get":

		logg("GET > $query");

		if (!$query) { error("GET['query'] is not defined for action '$action'"); }

		// Query
		$queryobj = json_decode($query,true);
		switch($queryobj["get"]) {

			// stations
			// TODO: status 304 if not modified
			// TODO: work in queues? list may become quite large...
			case "stations":
				$filename = "../json/stations.json";
				$json["data"] = fr($filename);
				if (!$json["data"]) { error("Error: file '$filename' not found"); }
				$json["data"] = json_decode($json["data"],true);
				if (!$json["data"]) { error("Error decoding json string: ". getJsonError(json_last_error())); }
				$json["info"]["last_update_time_ms"] = filemtime($filename)*1000; // TODO: More info?
				$json["info"]["desc"] = $queryobj["get"];
				$jsons = gzencode(json_encode($json));
				header('Content-Encoding: gzip');
				echo $jsons;
				break;

			// station_info
			case "station_info":
				if (!$queryobj["station_id"]) { error("Error: 'station_id' not defined for get:station_info"); }
				$filename = "../json/station_info.".$queryobj["station_id"].".json";
				if (@filemtime($filename)<time()-30) { // refresh file every xx secs
					if (!$queryobj["station_port"]) { $queryobj["station_port"] = 80; }
					// $id3_reader_url = "http://". $_SERVER['HTTP_HOST'] ."/icerrr/php/tests/test-readid3.php?q="; // TODO: rejh.nl only opens fsock on port 80
					$id3_reader_url = $cfg["icerrr_local_url"] . "php/tests/test-readid3.php?q=";
					if ($queryobj["station_port"]==80) {
						$id3_reader_url = $cfg["icerrr_remote_url"] . "php/tests/test-readid3.php?q=";
					}
					// "http://www.rejh.nl/icerrr/php/tests/test-readid3.php?q=";
                    $id3_reader_q_a = array(
                        "station_id" => $queryobj["station_id"],
                        "host" => $queryobj["station_host"],
                        "port" => $queryobj["station_port"],
                        "path" => $queryobj["station_path"],
                        "user" => $queryobj["station_user"],
                        "pass" => $queryobj["station_pass"]
                    );
					$id3_reader_q = urlencode(json_encode($id3_reader_q_a));
					// retry this a couple of times..
					$fgjson = false;
					$whilenum = 0;
					while(!$fgjson) {
						logg(" >> Read stream... $whilenum $id3_reader_url");
						if ($whilenum>=1) { break; }
						$whilenum++;
						$fg = fg($id3_reader_url.$id3_reader_q);
						$fgjson = @json_decode($fg,true);
					}
					// logg($id3_reader_q);
					if (!$fg) {
						error("Error talking to id3 reader: '".$id3_reader_url.$id3_reader_q."', '$fg', ". json_encode(error_get_last()));
					}
					$fw = fw($filename,$fg);
					if (!$fw) { error("Could not write $filename"); }
				}
				// sleep(1);
				$fr = fr($filename);
				if (!$fr) { error("Error: could not read file '$filename'"); } // TODO: Generate file :D
				$json["data"] = json_decode($fr,true);
				if (!$json["data"]) { logg(" >> ".$fr); error("Error: could not parse json: '$filename'"); }
				//$json["data"] = json_decode($json["data"],true);

				if ($json["data"]["nowplaying"]) {

					$npexpl = explode("-",$json["data"]["nowplaying"],2);
					$npartist = trim($npexpl[0]);
					$nptitle = trim($npexpl[1]);

					if (strpos($nptitle,"|")>0) {
						$nptitle = trim(substr($nptitle,0,strpos($nptitle,"|")-1));
					}

					$npartist = urlencode(strtolower(trim($npartist)));
					$nptitle = urlencode(strtolower(trim($nptitle)));

					$echonest_requrl = $cfg["echonest_apiurl"] ."song/search?"
						."api_key=". $cfg["echonest_apikey"]
						."&format=json&results=5"
						."&artist={$npartist}&title={$nptitle}"
						."&bucket=id:spotify&bucket=tracks"
						;

					$enjsons = fg($echonest_requrl);
					if ($enjsons && $npartist && $nptitle) {

						// Song search..
						$enjson = json_decode($enjsons,true);
						if ($enjson && $enjson["response"]["status"]["message"] == "Success") {

							// Artist, title
							if (count($enjson["response"]["songs"])>0) {
								$artist = $enjson["response"]["songs"][0]["artist_name"];
								$title = $enjson["response"]["songs"][0]["title"];
								$json["data"]["npartist"] = $artist;
								$json["data"]["nptitle"] = $title;
							}

							// Track id?
							for ($i=0; $i<count($enjson["response"]["songs"]); $i++) {
								if (count($enjson["response"]["songs"][$i]["tracks"])>0) {
									// release_image
									$trackid = $enjson["response"]["songs"][$i]["tracks"][0]["id"];
									break;
								}
							}

						}
						$json["data"]["npechores"] = ($enjson["response"]["status"]["message"] == "Success" && $json["data"]["npartist"] && $json["data"]["nptitle"]);

					}

					// Debug
					$json["data"]["npechoreq_songsearch"] = $echonest_requrl;

				}

				$json["info"]["last_update_time_ms"] = filemtime($filename)*1000; // TODO: More info?
				$json["info"]["last_update_timeago_sec"] = time() - filemtime($filename);
				$json["info"]["last_update_date"] = date("Y-m-d H:i:s",filemtime($filename));
				$json["info"]["desc"] = $queryobj["get"];
				$json["info"]["id3_reader_url"] = $id3_reader_url.$id3_reader_q;
				$jsons = json_encode($json);
				$jsons = gzencode($jsons);
				header('Content-Encoding: gzip');
				echo $jsons;
				break;

			// station_nowplaying // TODO: deprecated?
			case "station_nowplaying":
				if (!$queryobj["station_id"]) { error("Error: 'station_id' not defined for get:station_info"); }
				$filename = "../json/station_nowplaying.".$queryobj["station_id"].".json";
				$json["data"] = fr($filename);
				if (!$json["data"]) { error("Error: file '$filename' not found"); } // TODO: Generate file :D
				//$json["data"] = json_decode($json["data"],true);
				$json["info"]["last_update_time_ms"] = filemtime($filename)*1000; // TODO: More info?
				$json["info"]["desc"] = $queryobj["get"];
				$jsons = gzencode(json_encode($json));
				header('Content-Encoding: gzip');
				echo $jsons;
				break;

			// checkforupdates
			case "checkforupdates":
				if (!$queryobj["lookup"]) { error("Error: 'lookup' not defined for get:checkforupdates"); }
				if (!$queryobj["last_update_time_ms"]) { error("Error: 'last_update_time_ms' not defined for get:checkforupdates"); }
				if (!file_exists("../json/".$queryobj["lookup"].".json")) { error("Error: invalid lookup '".$queryobj["lookup"]."', file not found"); }
				$filemtime_ms = @filemtime("../json/".$queryobj["lookup"].".json")*1000;
				if ($queryobj["last_update_time_ms"] < filemtime_ms) { $json["data"] = 1; } // yes you need to update
				else { $json["data"] = 0; }
				$json["info"]["desc"] = $queryobj["get"];
				$jsons = json_encode($json);
				echo $json;
				break;

			// parse_playlist
			// - get stream(s) from pls and m3u files
			case "parse_playlist":
				if (!$queryobj["url"]) { error("Error: 'url' not defined for get:".$queryobj["get"]); }
				$fg = fg($queryobj["url"]);
				if (!$fg) { error("Error: Could not load url: '". $queryobj["url"] ."', {$php_errormsg}"); }
				$bgn = 0; $end = -1; $len = -1;
				$bgn = strpos($fg,"http");
				if ($bgn===FALSE) { error("Error: could not parse pls/m3u: '". $queryobj["url"] ."'"); }
				$end = strpos($fg,"\n",$bgn);
				$len = $end - $bgn;
				if ($end===FALSE) { $len = strlen($fg)-$bgn; }
				$url = trim( substr($fg,$bgn,$len), " \n\r\t");
				$json["data"] = $url;
				$json["info"] = array();
				$jsons = gzencode(json_encode($json));
				header('Content-Encoding: gzip');
				echo $jsons;
				break;

			// strings
			case "strings":
				error("Not implemented yet"); // TODO: todo
				break;

			// -- MESSAGES

			case "messages":
				$msgsStr = fg("data/messages.json");
				if (!msgsStr) { error("Could not load messages.json"); }
				$msgs = json_decode($msgsStr,true);
				if (!$msgs) { error("Could not parse messages.json: {$msgs}"); }
				$json["info"] = array();
				$json["data"] = $msgs;
				$jsons = gzencode(json_encode($json));
				header('Content-Encoding: gzip');
				echo $jsons;
				break;

			// -- EXTERNAL: Google Custom Search api

			// Google Image Search (via Google Custom Search API)
			case "gcisearch":
				// check cache
				$cachekey = $queryobj["search"]."-size".$queryobj["imgSize"];
				$cachedata = getGcisearchCache($cachekey);
				// New search..
				if (!$cachedata) {
                    logg(" > New search...");
					// Build url
					$url = $cfg["gcs_url"]
						."q=". rawurlencode($queryobj["search"])
						."&searchType=image"
						."&key=". $cfg["gcs_server_key"]
						."&cx=". $cfg["gcs_cefs"]
						;
					// userIP if available
					if ($queryobj["imgSize"]) {
						$url .= "&imgSize=". $queryobj["imgSize"];
					}
					if ($_SERVER['REMOTE_ADDR']!="127.0.0.1") {
						$url .= "&userIp=". $_SERVER['REMOTE_ADDR'];
					}
					// get + parse json
					$fg = fg($url);
					$fgjson = json_decode($fg,true);
					$json["data"] = $fgjson;
					// save to cache..
					saveGcisearchCache($cachekey, $json["data"]);
				}
				// use cache..
				else {
                    logg(" > Use cache...");
					$json["data"] = $cachedata;
					$json["info"]["cached"] = true;
				}
				$json["info"]["url"] = $url;
				$json["info"]["cachekey"] = $cachekey;
				$jsons = json_encode($jsons);
				$jsons = gzencode(json_encode($json));
				header('Content-Encoding: gzip');
				echo $jsons;
				break;

			// -- EXTERNAL: Dirble api

			// search // TODO: deprecated?
			case "search_dirble":
				if (!$queryobj["search"]) { error("Error: 'search' not defined for get:{$queryobj['get']}"); }
				$dirble_url = "http://api.dirble.com/v1/search/apikey/{$cfg['dirble_apikey']}/search/";
				$dirble_query = rawurlencode("{$queryobj['search']}");
				$data = getdirblecache($queryobj["get"], $dirble_query);
				if (!$data) {
					$fg = fg($dirble_url.$dirble_query);
					if (!$fg) { error("Error running search on Dirble: '". $dirble_url.$dirble_query."'"); }
					$data = json_decode($fg,true);
					savedirblecache($queryobj["get"], $dirble_query, $data);
				}
				$json["data"] = $data;
				$json["info"] = array();
				// TODO: catch errors
				$jsons = gzencode(json_encode($json));
				header('Content-Encoding: gzip');
				echo $jsons;
				break;

			// search v2 -> more
			case "search_dirble_v2":

				// Prep
				if (!$queryobj["search"]) { error("Error: 'search' not defined for get:{$queryobj['get']}"); }
				$dirble_url = "http://api.dirble.com/v2/search/";
				$dirble_query = rawurlencode("{$queryobj['search']}");

				// Cache or new
				$searchresults = getdirblecache($queryobj["get"], $dirble_query);
				if (!$searchresults) {

					$drible_url = $dirble_url . $dirble_query . "?token={$cfg['dirble_apikey']}";
					$searchresults = array();
					$results_primary = array();
					$results_secundary = array();
					$page = 1;

					// First query
					$fg = fg($drible_url);
					if (!$fg) { error("Error running search on Dirble: '". $dirble_url.$dirble_query."'"); }
					$res = json_decode($fg,true);
					if (!$res) { $res = array(); }

					// Add results
					for ($i=0; $i<count($res); $i++) {
						if (!$res[$i]) { continue; }
						$namelower = strtolower($res[$i]["name"]);
						$querylower = strtolower($queryobj['search']);

						// Primary/secundary results..
						if (strpos($namelower,$querylower)!==FALSE) {
							$results_primary[] = $res[$i];
						} else {
							$results_secundary[] = $res[$i];
						}
					}

					// Keep getting results..
					while(count($res)>0) {

						$page++;
						$fg = fg($drible_url . "&page={$page}");
						if (!$fg) { error("Error running search on Dirble: '". $dirble_url.$dirble_query."'"); }
						$res = json_decode($fg,true);
						if (!$res) { $res = array(); }

						//header("Content-Type: text/plain");
						//print_r($res);
						// echo "oi<br>";

						// Add results
						for ($i=0; $i<count($res); $i++) {
							if (!$res[$i]) { continue; }
							$namelower = strtolower($res[$i]["name"]);
							$querylower = strtolower($queryobj['search']);

							// Primary/secundary results..
							if (strpos($namelower,$querylower)!==FALSE) {
								$results_primary[] = $res[$i];
							} else {
								$results_secundary[] = $res[$i];
							}
						}

						// Merge
						$searchresults = array_merge($results_primary,$results_secundary);

						// Count
						if (count($searchresults)>=64) {
							break;
						}

					}
					savedirblecache($queryobj["get"], $dirble_query, $searchresults);
				}
				// Build response
				$json["data"] = $searchresults;
				$json["info"] = array();
				// TODO: catch errors
				$jsons = gzencode(json_encode($json));
				header('Content-Encoding: gzip');
				echo $jsons;
				break;

			case "nowplaying_dirble":
				if (!$queryobj["dirble_id"]) { error("Error: 'dirble_id' not defined for get:{$queryobj['get']}"); }
				$dirble_url = "http://api.dirble.com/v1/station/apikey/{$cfg['dirble_apikey']}/id/";
				$dirble_query = rawurlencode("{$queryobj['dirble_id']}");
				$data = getdirblecache($queryobj["get"], $dirble_query);
				if (!$data) {
					$fg = fg($dirble_url.$dirble_query);
					if (!$fg) { error("Error running query on Dirble: '". $dirble_url.$dirble_query."'"); }
					$data = json_decode($fg,true);
					savedirblecache($queryobj["get"], $dirble_query, $data);
				}
				$json["data"] = $data;
				$json["info"] = array();
				// TODO: catch errors
				$jsons = gzencode(json_encode($json));
				header('Content-Encoding: gzip');
				echo $jsons;
				break;

			case "nowplaying_dirble_v2":
				if (!$queryobj["dirble_id"]) { error("Error: 'dirble_id' not defined for get:{$queryobj['get']}"); }
				$dirble_url = "http://api.dirble.com/v2/station/id/";
				$dirble_query = rawurlencode("{$queryobj['dirble_id']}");
				$data = getdirblecache($queryobj["get"], $dirble_query);
				if (!$data) {
					$fg = fg($dirble_url.$dirble_query);
					if (!$fg) { error("Error running query on Dirble: '". $dirble_url.$dirble_query."'"); }
					$data = json_decode($fg,true);
					savedirblecache($queryobj["get"], $dirble_query, $data);
				}
				$json["data"] = $data;
				$json["info"] = array();
				// TODO: catch errors
				$jsons = gzencode(json_encode($json));
				header('Content-Encoding: gzip');
				echo $jsons;
				break;

			case "station_info_dirble_v2":
				if (!$queryobj["dirble_id"]) { error("Error: 'dirble_id' not defined for get:{$queryobj['get']}"); }
				$dirble_url = "http://api.dirble.com/v2/station/";
				$dirble_query = rawurlencode("{$queryobj['dirble_id']}");
				$dirble_url = $dirble_url . $dirble_query . "?token={$cfg['dirble_apikey']}";
				$data = getdirblecache($queryobj["get"], $dirble_query);
				if (!$data) {
					$fg = fg($dirble_url);
					if (!$fg) { error("Error running query on Dirble: '". $dirble_url ."'"); }
					$data = json_decode($fg,true);
					savedirblecache($queryobj["get"], $dirble_query, $data);
				}
				$json["data"] = $data;
				$json["info"] = array();
				// TODO: catch errors
				$jsons = gzencode(json_encode($json));
				header('Content-Encoding: gzip');
				echo $jsons;
				break;

			case "playlist_dirble_v2":
				// error("Not implemented yet :(");
				if (!$queryobj["dirble_id"]) { error("Error: 'dirble_id' not defined for get:{$queryobj['get']}"); }
				$dirble_url = "http://api.dirble.com/v2/station/";
				$dirble_query = rawurlencode("{$queryobj['dirble_id']}") . "/song_history";
				$dirble_url = $dirble_url . $dirble_query . "?token={$cfg['dirble_apikey']}";
				$data = getdirblecache($queryobj["get"], $dirble_query);
				if (!$data) {
					$fg = fg($dirble_url);
					if (!$fg) { error("Error running query on Dirble: '". $dirble_url ."'"); }
					$data = json_decode($fg,true);
					savedirblecache($queryobj["get"], $dirble_query, $data);
				}
				$json["data"] = $data;
				$json["info"] = array();
				// TODO: catch errors
				$jsons = gzencode(json_encode($json));
				header('Content-Encoding: gzip');
				echo $jsons;
				break;

			case "register_device":
				if (!$queryobj["id"]) { error("Error: !queryobj[id]"); }
				$id = $queryobj["id"];
				$jsons = fr("data/deviceids.json");
				$json = json_decode($jsons,true);
				if (!$json) { $json = array(); }
				if (!in_array($id,$json)) {
					$json[] = $id;
					$jsons = json_encode($json);
					$fw = fw("data/deviceids.json",$jsons);
					if (!$fw) { error("Error: could not write json file"); }
				}
				$json = array(); // purge json for response
				$json["data"] = array("saved"=>$id);
				$json["info"] = array();
				$jsons = gzencode(json_encode($json));
				header('Content-Encoding: gzip');
				echo $jsons;
				break;

			case "analytics":
				if (!$queryobj["id"]) { error("Error: !queryobj[id]"); }
				$id = $queryobj["id"];
				$device_model = $queryobj["device_model"];
				$device_platform = $queryobj["device_platform"];
				$app_version = $queryobj["app_version"];
				$time = time();

				$datas = fr("data/analytics.json");
				$data = json_decode($datas,true);
				if (!$data) { $data = array(); }
				$data[] = array(
					"id"=>$id,
					"device_model"=>$device_model,
					"device_platform"=>$device_platform,
					"app_version"=>$app_version,
					"time"=>time(),
					"datetime"=>date("Y-m-d H:i:s")
				);
				$datas = json_encode($data);
				$fw = fw("data/analytics.json",$datas);

				$json = array();
				$json["data"] = array("saved"=>$id,"time"=>time());
				$json["info"] = array();
				$jsons = gzencode(json_encode($json));
				header('Content-Encoding: gzip');
				echo $jsons;
				break;

			// default
			default:
				error("Error in query: '$query', {$queryobj['get']}, {$queryobj}");
				break;


		}
		break; // <-- Important stuff

	// POST ...
	case "post":

		logg("POST > $query");

		if (!$query) { error("GET['query'] is not defined for action '$action'"); }
		if (!$apikey) { error("API key is not provided"); }
		// TODO: this is such a bad security feat.. :()
		if (strpos($apikey,"REJH_ICERRR_APIKEY-")===FALSE) { error("API key is invalid"); } // TODO: Haha lol I call this a api key? XD

		// Query
		$queryobj = json_decode($query,true);
		switch($queryobj["post"]) {

			case "log":
				$id = $_POST["log_id"];
				$html = $_POST["log_html"];
				$text = $_POST["log_text"];
				if (!$id) { error("Error: !post[log_id]"); }
				if (!$html) { error("Error: !post[log_html]"); }
				$filenamehtml = "logs_users/html/{$id}.html";
				$filenametext = "logs_users/text/{$id}.txt";
				$html = "<html><style>body{font-family:sans-serif;font-size:10pt;}</style><body>{$html}</body></html>";
				$fw = fw($filenamehtml,$html);
				$fw = fw($filenametext,$text);
				if (!$fw) { error("Error: Could not write '$filename'"); }
				$json["data"] = json_decode('{"post":"ok"}',true);
				$json["info"] = array();
				$jsons = json_encode($json);
				logg($jsons);
				echo $jsons;
				sendEmail_log("https://www.rejh.nl/icerrr/api/{$filenamehtml}");
				break;

			case "stations":
				$id = $_POST["id"];
				$stations = $_POST["stations"];
				if (!$id || !$stations) { error("Error: !post[id] || !post[stations]"); }
				$filename = "stations_users/stations_{$id}.json";
				$fw = fw($filename,$stations);
				if (!$fw) { error("Error: Could not write '$filename'"); }
				$json["data"] = json_decode('{"post":"ok"}',true);
				$json["info"] = array();
				$jsons = json_encode($json);
				logg($jsons);
				echo $jsons;
				break;

			case "station":
				$id = $_POST["id"];
				$station = $_POST["station"];
				if (!$id || !$stations) { error("Error: !post[id] || !post[station]"); }
				$filename = "station_users/station_{$id}.json";
				$fw = fw($filename,$station);
				if (!$fw) { error("Error: Could not write '$filename'"); }
				$json["data"] = json_decode('{"post":"ok"}',true);
				$json["info"] = array();
				$jsons = json_encode($json);
				logg($jsons);
				echo $jsons;
				break;

			case "station_icon": // NOTE: does not return json, it returns http responses! // TODO: yes?

				$filename = $_FILES['file']['name'];
				$filepath = "../static/uploaded/{$filename}";
				$filetmp = $_FILES['file']['tmp_name'];

                /*
                // redundant check. this is easily faked..
				$postkey = $_POST["key"];
				$getkey = $queryobj["key"];
				if ($postkey!=$getkey) {
					http_response_code(403);
					error("Access denied'");
				}
                /**/

				$devicess = fr("data/deviceids.json");
				$devices = json_decode($devicess,true);
				if (!in_array($queryobj["device"],$devices)) {
					http_response_code(403);
					error("Access denied'");
				}

				$moved = move_uploaded_file($filetmp, $filepath);
				if (!$moved) {
					http_response_code(500);
					error("Could not move file '{$filetmp}' to '{$filepath}'");
				} else {
					http_response_code(200);
					echo json_encode(array("info"=>array(), "data"=>array("post"=>"ok", "filename"=>$filename)));
				}
				break;

			// default
			default:
				error("Error in query: '$query', {$queryobj['post']}, {$queryobj}");
				die();

		}
		break;

	// Default
	default:
		error("Action '{$action}' is not supported");
		die();

}


















?>
